src/hotspot/share/opto/bytecodeInfo.cpp

```diff
@@ -315,7 +315,13 @@ bool InlineTree::should_not_inline(ciMethod* callee_method, ciMethod* caller_met
       int invoke_count     = caller_method->interpreter_invocation_count();
       assert(invoke_count != 0, "require invocation count greater than zero");
       double freq = (double)call_site_count / (double)invoke_count;
-      double min_freq = MAX2(MinInlineFrequencyRatio, 1.0 / CompilationPolicy::min_invocations());
+      double min_freq = 0.0;
+      int cp_min_inv = CompilationPolicy::min_invocations();
+      if (cp_min_inv == 0) {
+        min_freq = MinInlineFrequencyRatio;
```

Comment:
Doesn't this need to be 1.0 or infinity to preserve the existing behavior?
However, I'm not convinced the existing behavior is correct.  If min_invocations decreases, shouldn't that make it easier to inline, not harder?  @veresov, do you agree?

Reply:
I think the existing behavior is correct - the goal is to compute a minimum call freq that is considered for inlining. `1.0/min_invocations()` means call site was reached at least once before the method was submitted for a C2 compile. 

The question is what should we do if min_invocations() is 0. That basically means that there was no profiling. I would probably just return false in that case, which would mean "we can't really say that's low frequency, because we don't have data". It's a meaningless mode of operation regardless though...

Reply:
>  think the existing behavior is correct - the goal is to compute a minimum call freq that is considered for inlining. 1.0/min_invocations() means call site was reached at least once before the method was submitted for a C2 compile.

Do we really need to take `1.0/min_invocations()` into account?  `InlineSmallCode` should handle cases like that.

Also `MinInlineFrequencyRatio ` is 0.0085 which corresponds to  `Tier4MinInvocationThreshold` value equal approximately  117. The default value is 600. The only way to change it is to use `CompileThreshold` or `CompileThresholdScaling` which nobody use in production but only in testing. So in production we use `MinInlineFrequencyRatio` as `min_freq`.  Why complicate the code.
 
An other issue. This is `should_not_inline()` method. Returning `false` means **inlining** called method. I was curious why we return `false` when we run with `-Xcomp` (see check `(UseInterpreter)` at line 301) ?
If we return `false` when `min_invocations()` is 0 it will be similar to `-Xcomp` mode.

Is it intentional to inline when we don't have profiling info?

Reply:
> > think the existing behavior is correct - the goal is to compute a minimum call freq that is considered for inlining. 1.0/min_invocations() means call site was reached at least once before the method was submitted for a C2 compile.
> 
> Do we really need to take `1.0/min_invocations()` into account? `InlineSmallCode` should handle cases like that.
> 

I'm not sure I understand the question. `InlineSmallCode` acts on already compiled methods. This logic analyzes call site frequencies.

> Also `MinInlineFrequencyRatio ` is 0.0085 which corresponds to `Tier4MinInvocationThreshold` value equal approximately 117. The default value is 600. The only way to change it is to use `CompileThreshold` or `CompileThresholdScaling` which nobody use in production but only in testing. So in production we use `MinInlineFrequencyRatio` as `min_freq`. Why complicate the code.

Can't exactly tell. But I'm sure I had a reason, probably some corner case exposed by testing. Probably something  didn't get inlined with very low thresholds.

> 
> An other issue. This is `should_not_inline()` method. Returning `false` means **inlining** called method. I was curious why we return `false` when we run with `-Xcomp` (see check `(UseInterpreter)` at line 301) ? If we return `false` when `min_invocations()` is 0 it will be similar to `-Xcomp` mode.
> 
> Is it intentional to inline when we don't have profiling info?

I think some tests run with -Xcomp and expect inlining to happen.

Reply:
So lowering CompileThreshold to below 117, means "compile sooner", but inline less, because should_not_inline will return true more often?

Reply:
Yes, the problem is that the tests still expect you to inline things even with small or 0 thresholds. Hence the `1/min_invocations()` floor.

Reply:
But this turns into should not inline for freq < infinity or freq < 1, or basically "always should not inline".  Wouldn't we want to use MIN2 instead of MAX2 for floor?

Reply:
Good question. I think what I originally meant it to be is the minimum measurable frequency. In that case MAX2 would work. On the other hand, it's not a scaled value. I think we can try removing it and leaving only `freq < MinInlineFrequencyRatio` and see which test fails and why. I, unfortunately, don't remember the details.